flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -147,6 +150,148 @@ static void processFirstRow(
 		out.collect(currentRow);
 	}
 
+	/**
+	 * Processes element to deduplicate on keys with row time semantic, sends current element if it is last
+	 * or first row, retracts previous element if needed.
+	 *
+	 * @param state                 state of function
+	 * @param currentRow            latest row received by deduplicate function
+	 * @param out                   underlying collector
+	 * @param generateUpdateBefore  flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert        flag to gennerate INSERT message or not
+	 * @param orderFunctionProvider provider that provides an order function to judge first or last
+	 */
+	public static void deduplicateOnRowTime(
+			ValueState<RowData> state,
+			RowData currentRow,
+			Collector<RowData> out,
+			boolean generateUpdateBefore,
+			boolean generateInsert,
+			DeduplicateOrderFunctionProvider orderFunctionProvider) throws Exception {
+
+		checkInsertOnly(currentRow);
+		RowData preRow = state.value();
+		if (!orderFunctionProvider.matches(preRow, currentRow)) {
+			return;
+		}
+
+		collectRetractResult(
+				generateUpdateBefore,
+				generateInsert,
+				preRow,
+				currentRow,
+				out,
+				null
+		);
+		state.update(currentRow);
+	}
+
+	/**
+	 * Processes element to deduplicate on keys with row time semantic, sends current element if it is last
+	 * or first row, retracts previous element if needed.
+	 *
+	 * @param state                 state of function
+	 * @param bufferedRows          latest row received by deduplicate function
+	 * @param serializer            serializer to serialize the data
+	 * @param out                   underlying collector
+	 * @param generateUpdateBefore  flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert        flag to gennerate INSERT message or not
+	 * @param orderFunctionProvider provider that provides an order function to judge first or last
+	 */
+	public static void miniBatchDeduplicateOnRowTime(
+			ValueState<RowData> state,
+			List<RowData> bufferedRows,
+			TypeSerializer<RowData> serializer,
+			Collector<RowData> out,
+			boolean generateUpdateBefore,
+			boolean generateInsert,
+			DeduplicateOrderFunctionProvider orderFunctionProvider) throws Exception {
```

Comment:
If this method can be moved in `RowTimeMiniBatchDeduplicateFunction`,  I think we don't need the special `DeduplicateOrderFunctionProvider` parameter then, because we have `keepLastRow` flag.